### PR TITLE
Optimize LoadAccount

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Separated getCode from loadAccount - code is only got from trie when it is needed.
Added fixes on esdt callbacks and return of values
Integrated new arwen

### CODE REVIEW

File: data/state/accountsDB.go [hunk 1]
```diff
@@ -87,6 +87,34 @@ func getNumCheckpoints(trie data.Trie) uint32 {
...
 }
 
+func (adb *AccountsDB) GetCode(account AccountHandler) []byte {
+	baseAcc, ok := account.(baseAccountHandler)
+	if !ok {
+		return nil
+	}
+
+	if len(baseAcc.GetCode()) != 0 {
+		return baseAcc.GetCode()
```
Comment (iulianpascalau) [inline data/state/accountsDB.go:97]: assuming the base account has a code inside and returning it in that case might lead to this scenario:
I have an account, I've put by mistake something in its code field and then I call AccountsDB.GetCode assuming I will get the "real" contract code. Instead I get the existing byte slice from the user account.

File: data/state/accountsDB.go [hunk 2]
```diff
@@ -505,10 +537,10 @@ func (adb *AccountsDB) LoadAccount(address []byte) (AccountHandler, error) {
 
 	baseAcc, ok := acnt.(baseAccountHandler)
 	if ok {
-		err = adb.loadCode(baseAcc)
-		if err != nil {
-			return nil, err
-		}
+		//err = adb.loadCode(baseAcc)
```
Comment (iulianpascalau) [inline data/state/accountsDB.go:540]: commented code

File: data/state/accountsDB.go [hunk 3]
```diff
@@ -564,10 +596,10 @@ func (adb *AccountsDB) GetExistingAccount(address []byte) (AccountHandler, error
...
```
Comment (iulianpascalau) [inline data/state/accountsDB.go:599]: commented code

File: data/state/baseAccount.go [hunk 4]
```diff
@@ -40,6 +42,10 @@ func (ba *baseAccount) DataTrieTracker() DataTrieTracker {
 	return ba.dataTrieTracker
 }
 
+func (ba *baseAccount) HasNewCode() bool {
```
Comment (iulianpascalau) [inline data/state/baseAccount.go:45]: missing comment

File: data/state/baseAccount.go [hunk 5]
```diff
@@ -22,6 +23,7 @@ func (ba *baseAccount) GetCode() []byte {
 
 // SetCode sets the actual code that needs to be run in the VM
 func (ba *baseAccount) SetCode(code []byte) {
+	ba.hasNewCode = true
```
Comment (iulianpascalau) [inline data/state/baseAccount.go:26]: ba.hasNewCode = !bytes.Equal(code, ba.code) ?

File: epochStart/bootstrap/disabled/disabledAccountsAdapter.go [hunk 6]
```diff
@@ -16,6 +16,10 @@ func NewAccountsAdapter() *accountsAdapter {
 	return &accountsAdapter{}
 }
 
+func (a *accountsAdapter) GetCode(_ state.AccountHandler) []byte {
```
Comment (iulianpascalau) [inline epochStart/bootstrap/disabled/disabledAccountsAdapter.go:19]: Missing comment

File: node/node.go [hunk 7]
```diff
@@ -13,7 +13,7 @@ import (
 	"sync/atomic"
 	"time"
 
-	"github.com/ElrondNetwork/elrond-go-logger"
+	logger "github.com/ElrondNetwork/elrond-go-logger"
```
Comment (iulianpascalau) [inline node/node.go:16]: unnecessary alias

File: node/txsimulator/wrappedAccountsDB.go [hunk 8]
```diff
@@ -24,6 +24,10 @@ func NewReadOnlyAccountsDB(accountsDB state.AccountsAdapter) (*readOnlyAccountsD
 	return &readOnlyAccountsDB{originalAccounts: accountsDB}, nil
 }
 
+func (w *readOnlyAccountsDB) GetCode(account state.AccountHandler) []byte {
```
Comment (iulianpascalau) [inline node/txsimulator/wrappedAccountsDB.go:27]: Missing comment

File: data/state/accountsDB.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply data/state/accountsDB.go:97]: Should I retrieve the code from the trie each time?

File: data/state/accountsDB.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply data/state/accountsDB.go:599]: Removed.

File: data/state/baseAccount.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply data/state/baseAccount.go:45]: Added.

File: data/state/baseAccount.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply data/state/baseAccount.go:26]: Done.

File: epochStart/bootstrap/disabled/disabledAccountsAdapter.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply epochStart/bootstrap/disabled/disabledAccountsAdapter.go:19]: Added.

File: node/txsimulator/wrappedAccountsDB.go (hunk 8 above)
Comment (BeniaminDrasovean) [reply node/txsimulator/wrappedAccountsDB.go:27]: Added.

File: node/node.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply node/node.go:16]: Removed.

File: data/state/accountsDB.go (hunk 1 above)
Comment (iulianpascalau) [reply data/state/accountsDB.go]: I think it is better without this sort of cache (we can do it in other parts of the code). In this way this function express the intent in a more straight forward way.

**Review by iulianpascalau — APPROVED** [review]

File: data/state/accountsDB.go (hunk 1 above)
Comment (sasurobert) [reply data/state/accountsDB.go]: you can get it - as we have extra gas cost for it now.

**Review by sasurobert — CHANGES_REQUESTED** [review]
I think you should delete GetCode from the UserAccountsHandler interface

**Review by sasurobert — DISMISSED** [review]

File: integrationTests/vm/arwen/arwenVM/arwenVM_test.go [hunk 9]
```diff
@@ -246,8 +246,8 @@ func TestMultipleTimesERC20BigIntInBatches(t *testing.T) {
 	if testing.Short() {
 		t.Skip("this is not a short test")
 	}
-	gasSchedule, _ := core.LoadGasScheduleConfig("../../../../cmd/node/config/gasSchedules/gasScheduleV2.toml")
-	deployAndExecuteERC20WithBigInt(t, 3, 1000, gasSchedule, "../testdata/erc20-c-03/wrc20_arwen.wasm", "transferToken", false)
+	//gasSchedule, _ := core.LoadGasScheduleConfig("../../../../cmd/node/config/gasSchedules/gasScheduleV2.toml")
```
Comment (iulianpascalau) [inline integrationTests/vm/arwen/arwenVM/arwenVM_test.go:249]: remove commented code

File: node/node.go [hunk 10]
```diff
@@ -1124,7 +1124,8 @@ func (n *Node) GetAccount(address string) (state.UserAccountHandler, error) {
 	accWrp, err := n.accounts.GetExistingAccount(addr)
 	if err != nil {
 		if err == state.ErrAccNotFound {
-			return state.NewUserAccount(addr)
+			newAcc, err := state.NewUserAccount(addr)
```
Comment (iulianpascalau) [inline node/node.go:1127]: return directly?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: integrationTests/vm/arwen/arwenVM/arwenVM_test.go (hunk 9 above)
Comment (sasurobert) [reply integrationTests/vm/arwen/arwenVM/arwenVM_test.go:249]: done

File: node/node.go (hunk 10 above)
Comment (sasurobert) [reply node/node.go]: done

**Review by iulianpascalau — APPROVED** [review]

File: data/state/export_test.go [hunk 11]
```diff
@@ -25,6 +25,10 @@ func (adb *AccountsDB) GetObsoleteHashes() map[string][][]byte {
 	return adb.obsoleteDataTrieHashes
 }
 
+func GetCode(account baseAccountHandler) []byte {
```
Comment (SebastianMarian) [inline data/state/export_test.go:28]: GetCodeHash?

Comment (sasurobert) [reply data/state/export_test.go:28]: this is an export function only. some stuff is mocked for tests only.

**Review by sasurobert — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

File: data/state/export_test.go (hunk 11 above)
Comment (SebastianMarian) [reply data/state/export_test.go:28]: Yes it is, but it returns a code hash and not the code, the name is misleading also for tests :)

**Review by LucianMincu — APPROVED** [review]
System tests passed.

**Pull request merged**